conf_file.cpp

```diff
@@ -16,268 +16,416 @@
  * limitations under the License.
  */
 
-#include "conf_file.h"
-
 #include <assert.h>
 #include <ctype.h>
-#include <stdio.h>
-#include <stdlib.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
 #include <string.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <unistd.h>
 
+#include "conf_file.h"
 #include "log.h"
 #include "util.h"
 
 struct config {
-    char *key;
-    char *value;
+    const char *key;
+    const char *value;
+    size_t key_len, value_len;
     struct config *next;
 };
 
 struct section {
-    char *name;
+    const char *name;
+    size_t len;
     struct config *configs;
-    struct config *current_config;
     struct section *next;
 };
 
+struct conffile {
+    int fd;
+    void *addr;
+    size_t len;
+    struct conffile *next;
+};
+
 ConfFile::~ConfFile()
 {
-    struct section *section;
-    struct config *config;
-
-    section = _sections;
-    while (section) {
-        struct section *tmp = section;
-
-        config = section->configs;
-        while (config) {
-            struct config *tmp_config = config;
-            free(config->key);
-            // config->value is on the same allocation as config->key, no need to free it
-            config = config->next;
-            free(tmp_config);
-        }
-
-        free(section->name);
-        section = section->next;
-        free(tmp);
-    }
+    release_all();
 }
 
-int ConfFile::parse_file()
+int ConfFile::parse(const char *filename)
 {
-    FILE *file;
-    char *entry = NULL;
-    int line = 0, ret = 0;
-    size_t n = 0;
-    ssize_t read;
+    int fd, ret = 0;
+    void *addr;
+    struct stat fstat;
+
+    assert(filename);
+
+    fd = open(filename, O_RDONLY | O_CLOEXEC);
+    if (fd < 0)
+        return -errno;
+
+    if (stat(filename, &fstat) < 0)
+        return -errno;
 
-    assert(_filename);
+    addr = mmap(0, (size_t)fstat.st_size, PROT_READ, MAP_PRIVATE, fd, 0);
+    if (addr == MAP_FAILED)
+        return -errno;
 
-    file = fopen(_filename, "re");
-    if (!file) {
-        log_error_errno(errno, "Could not open conf file '%s' (%m)", _filename);
-        return -EIO;
+    _files = new conffile{fd, addr, (size_t)fstat.st_size, _files};
+
+    ret = _parse_file((char *)addr, (size_t)fstat.st_size, filename);
+    if (ret < 0) {
+        log_error("Conf file parcially parsed. Configuration keys defined before the error were "
+                  "included");
```

Comment:
I am just logging what has happened here and I am returning an error code.
I decided to print this because if you are parsing multiple files and the last one fails, some data from last file will be added anyway. I don't want people to assume that only data from previous files that were successfully parsed were added.